src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala

```diff
@@ -446,6 +447,26 @@ trait V2Formats {
           )
         }
       }
+  }.map(addHealthCheckPortIndexIfNecessary)
+
+  /**
+    * Ensure backwards compatibility by adding portIndex to health checks when necessary.
+    *
+    * In the past, healthCheck.portIndex was required and had a default value 0. When we introduced healthCheck.port, we
+    * made it optional (also with ip-per-container in mind) and we have to re-add it in cases where it makes sense.
+    */
+  private[this] def addHealthCheckPortIndexIfNecessary(v2App: V2AppDefinition): V2AppDefinition = {
+    val hasPortMappings = v2App.container.exists(_.docker.exists(_.portMappings.exists(_.nonEmpty)))
+    val portIndexesMakeSense = v2App.ports.nonEmpty || hasPortMappings
+    if (portIndexesMakeSense) {
+      v2App.copy(healthChecks = v2App.healthChecks.map { healthCheck =>
+        if (healthCheck.port.isEmpty && healthCheck.portIndex.isEmpty && healthCheck.protocol != Protocol.COMMAND) {
```

Comment:
You mean like this?

```scala

  /**
    * Ensure backwards compatibility by adding portIndex to health checks when necessary.
    *
    * In the past, healthCheck.portIndex was required and had a default value 0. When we introduced healthCheck.port, we
    * made it optional (also with ip-per-container in mind) and we have to re-add it in cases where it makes sense.
    */
  private[this] def addHealthCheckPortIndexIfNecessary(v2App: V2AppDefinition): V2AppDefinition = {
    val hasPortMappings = v2App.container.exists(_.docker.exists(_.portMappings.exists(_.nonEmpty)))
    val portIndexesMakeSense = v2App.ports.nonEmpty || hasPortMappings
    v2App.copy(healthChecks = v2App.healthChecks.map { healthCheck =>
        def needsDefaultPortIndex = healthCheck.port.isEmpty && healthCheck.portIndex.isEmpty && healthCheck.protocol != Protocol.COMMAND
        if (portIndexesMakeSense && needsDefaultPortIndex) healthCheck.copy(portIndex = Some(0))
        else healthCheck
    })
  }
```

Slightly less efficient but more readable, I like it.